jira/resilientsession.py

```diff
@@ -1,119 +1,262 @@
+    """Parse a Jira Error message from the Response.
+
+    https://developer.atlassian.com/cloud/jira/platform/rest/v2/intro/#status-codes
+
+    Args:
+        resp (Response): The Jira API request's response.
+
+    Returns:
+        str: The error message parsed from the Response. An empty string if no error.
+    """
+    resp_data: Dict[str, Any] = {}  # json parsed from the response
+    parsed_error = ""  # error message parsed from the response
+
+    if resp.status_code == 403 and "x-authentication-denied-reason" in resp.headers:
+        parsed_error = resp.headers["x-authentication-denied-reason"]
+    elif resp.text:
+        try:
+            resp_data = resp.json()
+        except ValueError:
+            parsed_error = resp.text
+
+    if "message" in resp_data:
+        # Jira 5.1 errors
+        parsed_error = resp_data["message"]
+    elif "errorMessages" in resp_data:
+        # Jira 5.0.x error messages sometimes come wrapped in this array
+        # Sometimes this is present but empty
+        error_messages = resp_data["errorMessages"]
+        if len(error_messages) > 0:
+            if isinstance(error_messages, (list, tuple)):
+                parsed_error = error_messages[0]
+            else:
+                parsed_error = error_messages
+    elif "errors" in resp_data:
+        resp_errors = resp_data["errors"]
+        if len(resp_errors) > 0 and isinstance(resp_errors, dict):
+            # Catching only 'errors' that are dict. See https://github.com/pycontribs/jira/issues/350
+            # Jira 6.x error messages are found in this array.
+            error_list = resp_errors.values()
+            parsed_error = ", ".join(error_list)
+
+    return parsed_error
 
 
 class ResilientSession(Session):
     """This class is supposed to retry requests that do return temporary errors.
 
-    At this moment it supports: 429
+    :py:meth:`__recoverable` handles all retry-able errors.
     """
 
-    def __init__(self, timeout=None):
-        self.max_retries = 3
-        self.max_retry_delay = 60
-        self.timeout = timeout
+    def __init__(self, timeout=None, max_retries: int = 3, max_retry_delay: int = 60):
+        """A Session subclass catered for the Jira API with exponential delaying retry.
+
+        Args:
+            timeout (Optional[int]): Timeout. Defaults to None.
+            max_retries (int): Max number of times to retry a request. Defaults to 3.
+            max_retry_delay (int): Max delay allowed between retries. Defaults to 60.
+        """
+        self.timeout = timeout  # TODO: Unused?
+        self.max_retries = max_retries
+        self.max_retry_delay = max_retry_delay
         super().__init__()
 
         # Indicate our preference for JSON to avoid https://bitbucket.org/bspeakmon/jira-python/issue/46 and https://jira.atlassian.com/browse/JRA-38551
         self.headers.update({"Accept": "application/json,*.*;q=0.9"})
 
+    def _jira_prepare(self, **_kwargs) -> dict:
```

Comment:
why _kwargs and not kwargs here? I might not be fully up to date on the standards here

Reply:
nah it just a bad variable name, I'll updated the names

(recall, `**variable_name` just takes any additional keyword arguments passed into the function that are not explicitly listed in the function definition)